### DISCUSSION TITLE
RC 101 API hooks broken.

### DISCUSSION THREAD

**Discussion by NecoHorne** [directus/directus#9453] [Archive]
RC 101 completely broke all my hooks and the updated documentation does not help. 

Old code:

```
'items.update': function ({ item, collection, database}){
            if (collection === 'articles'){
                database.from('articles').where({
                    id: item
                }).select('*').then(async (rows) => {
                    for (const row of rows) {
                            let articleIndex = await getArticle(row, database);

                            if (row['status'] === 'published'){
                                //update the article on the news service
                                updateArticle(articleIndex);
                            }else if (row['status'] === 'draft'){
                                //remove the article from the news service
                                deleteArticle(articleIndex.id);
                            }else if (row['status'] === 'archived'){
                                //remove the article from the news service
                                deleteArticle(articleIndex.id);
                            }
                    }
                }).catch((error) => {
                    console.log(error)
                });
            }
        },
```

How would I do this with the new code? The new API documentation is seriously lacking, really frustrated with this update. 

Do I now use `filter()` instead of the `.before` ? Do I use `action()` instead of the normal and `.after`? 

for `action('items.update')` the meta contains an array of keys, does this mean it only fires once now instead of for each update? I am so confused by this change.

**Comment by nickrum** [comment]
> The new API documentation is seriously lacking, really frustrated with this update.

What exactly is lacking in your opinion?

> Do I now use `filter()` instead of the `.before` ? Do I use `action()` instead of the normal and `.after`?

This is exactly the type of confusion we were trying to clarify. Yes, filters are the new `.before` events, actions are the new no-`.before` events. Everything that had a `.before` but didn't modify the dataflow or that had an `.after` is now an init event. This is explained in the docs when the four event types are introduced.

> for `action('items.update')` the meta contains an array of keys, does this mean it only fires once now instead of for each update?

The old `item` property of the `items.update` had always been an array of keys because you can update multiple items with a single API call. We just renamed it to make this more obvious.

> How would I do this with the new code?

Something like this should work:

```js
module.exports = function registerHook({ action }) {
    action('items.update', ({ keys, collection }, { database }) => {
        if (collection === 'articles'){
                database.from('articles').where({
                    id: keys
                }).select('*').then(async (rows) => {
                    for (const row of rows) {
                            let articleIndex = await getArticle(row, database);

                            if (row['status'] === 'published'){
                                //update the article on the news service
                                updateArticle(articleIndex);
                            }else if (row['status'] === 'draft'){
                                //remove the article from the news service
                                deleteArticle(articleIndex.id);
                            }else if (row['status'] === 'archived'){
                                //remove the article from the news service
                                deleteArticle(articleIndex.id);
                            }
                    }
                }).catch((error) => {
                    console.log(error)
                });
            }
    );
};
```

**Comment by NecoHorne** [comment]
Thanks, that makes more sense. The docs don't explain it like that, the main confusion came in with the action, filter, and init, Sorry I was a bit dramatic in that post, the change caught me off guard my project utilizes many, many hooks in very different ways, just used to how the old hooks work.

**Reply by rijkvanzanten** [reply]
Any recommendation on how to clarify the difference between those hooks a little better? We can update it straight away 🙂

**Reply by Fusseldieb** [reply]
@rijkvanzanten I believe including the following (from answer above) into the docs would be a huge help for those who updated mindlessly and broke their stuff (absolutely not me), into a blue info box:
> Filters are the new .before events, actions are the new no-.before events. Everything that had a .before but didn't modify the dataflow or that had an .after is now an init event.

**Reply by rijkvanzanten** [reply]
@jaycammarano just updated that page to hopefully make all this way more clear. Lemme know what you think!

**Comment by jawngee** [comment]
Went through this as well.  I think it boils down to maybe giving an example for each type of hook and the parameters to the callback for the hook.  For example, the documentation has this:

`The first parameter of the action register function is the event name. The second argument is an event-specific meta object.`

But that is kind of confusing and doesn't (in my mind), line up with the actual use of the hook:

```js
action('files.upload', async ({ key, payload }, { schema, accountability }) => { ... });
```

The first argument to the callback is actually the meta object.

**Reply by nickrum** [reply]
Yes, you are absolutely right. The descriptions of the register function parameters are clearly wrong. Should be fixed by #9489.
